lib/containers/k8s.pm

```diff
@@ -27,67 +27,110 @@ our @EXPORT = qw(install_k3s uninstall_k3s install_kubectl install_helm install_
-        script_retry("curl -sfL https://get.k3s.io  -o install_k3s.sh", timeout => 180, delay => 60, retry => 3);
-        assert_script_run("INSTALL_K3S_SKIP_START=true sh install_k3s.sh $disables", timeout => 300);
-        script_run("rm -f install_k3s.sh");
-    }
+sub kubectl {
+    return "kubectl @_";
+}
 
-    if (get_var('REGISTRY')) {
-        script_run("mkdir -p /etc/rancher/k3s");
-        my $registry = registry_url();
-        assert_script_run "curl " . data_url('containers/registries.yaml') . " -o /etc/rancher/k3s/registries.yaml";
-        file_content_replace("/etc/rancher/k3s/registries.yaml", REGISTRY => $registry);
-    }
-    systemctl('start k3s');
-    script_retry("test -e /etc/rancher/k3s/k3s.yaml", delay => 20, retry => 10);
-    systemctl('is-active k3s');
+sub check_k3s {
     assert_script_run('k3s -v');
     assert_script_run('uname -a');
     assert_script_run("k3s kubectl get node");
+    script_run('kubectl config get-clusters');
+    script_run('kubectl config get-contexts');
+    script_run('kubectl config get-users');
+    script_run('kubectl config view');
     validate_script_output_retry("k3s kubectl get node", qr/ Ready /, retry => 6, delay => 15, timeout => 90);
-    script_run("mkdir -p ~/.kube");
-    script_run("rm -f ~/.kube/config");
-    assert_script_run("ln -s /etc/rancher/k3s/k3s.yaml ~/.kube/config");
-    sleep 60;
+    assert_script_run("k3s kubectl get node");
+    script_run('kubectl config get-clusters');
+    script_run('kubectl config get-contexts');
+    script_run('kubectl config get-users');
+    script_run('kubectl config view');
     # Await k3s to be ready and exists the default service account
     script_retry("kubectl get serviceaccount default -o name", timeout => 120, delay => 60, retry => 3);
     # Await k3s to be ready and the api is accessible
     script_retry("kubectl get namespaces", timeout => 120, delay => 60, retry => 3);
-    script_retry("kubectl get pods --all-namespaces | grep -E 'Running|Completed'", timeout => 120, delay => 60, retry => 10);
+    script_retry("kubectl get pods --all-namespaces", timeout => 120, delay => 60, retry => 10);
     record_info("k3s api resources", script_output("kubectl api-resources"));
     assert_script_run("kubectl auth can-i 'create' 'pods'");
     assert_script_run("kubectl auth can-i 'create' 'deployments'");
     record_info('k3s', "k3s version " . script_output("k3s --version") . " installed");
     record_info('kubectl version', script_output('kubectl version --short'));
 }
 
+sub ensure_k3s_start {
+    systemctl('start k3s');
+    systemctl('is-active k3s');
+}
+
+sub set_custom_registry {
+    return unless get_var('REGISTRY');
+
+    script_run("mkdir -p /etc/rancher/k3s");
+    my $registry = registry_url();
+    assert_script_run "curl " . data_url('containers/registries.yaml') . " -o /etc/rancher/k3s/registries.yaml";
```

Comment:
I think this file is giving me issues because the mirror is not setup for only docker.io, so registry.suse.de fails.  It also pollutes the mirror if we use other registries like SUSE's and openSUSE's.

Reply:
I am not fully getting the problem. Can you share with a link to the failure, please?
Is it a search or pull problem?

Reply:
Ah, I have checked the wrong file in `data` folder. Well, it should pull images from our registry only. Are you missing an image that should be there? If so you need to pull it to our test registry

Reply:
> I am not fully getting the problem. Can you share with a link to the failure, please? Is it a search or pull problem?

[This file](https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/data/containers/registries.yaml) should be setup like this according to [documentation](https://docs.k3s.io/installation/private-registry)

```
mirrors:
  docker.io:
    endpoint:
      - "http://REGISTRY"
```

The way is currently setup makes any module to use the mirror registry for all images, even private images in registry.suse.de.  And not only private images but also publicly available images like in registry.opensuse.org that don't need such mirror.

Reply:
ah, I see. Do you then suggest adding mirrors for opensuse/suse.{de,com} the same as for docker.io?

Reply:
I cannot remember whether we reached the pull count for docker.io or for others as well in the pas

Reply:
> ah, I see. Do you then suggest adding mirrors for opensuse/suse.{de,com} the same as for docker.io?

Those don't need mirror.  I think the above snippet should work.

Reply:
yup, should do just fine, I will add it there. Thanks!

Reply:
Yup, the custom registry is only needed on docker.io because of the rate-limitation.